scio-extra/src/main/scala/com/spotify/scio/extra/voyager/VoyagerUri.scala

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2023 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.spotify.scio.extra.voyager
+
+import com.spotify.scio.coders.Coder
+import com.spotify.scio.util.{RemoteFileUtil, ScioUtil}
+import com.spotify.voyager.jni.Index
+import com.spotify.voyager.jni.Index.{SpaceType, StorageDataType}
+import org.apache.beam.sdk.options.PipelineOptions
+import org.slf4j.LoggerFactory
+
+import java.io.File
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Path, Paths}
+import scala.collection.mutable
+
+/**
+ * Represents the base URI for a voyager index, either on a local or a remote file system. For
+ * remote file systems, the `path` should be in the form 'scheme://<bucket>/<path>/'. For local
+ * files, it should be in the form '/<path>/'. The `path` specified represents the directory where
+ * the `index.hnsw` and `names.json` are.
+ */
+trait VoyagerUri extends Serializable {
+  val logger = LoggerFactory.getLogger(this.getClass)
+  val path: String
+  private[voyager] def getReader(
+    distanceMeasure: VoyagerDistanceMeasure,
+    storageType: VoyagerStorageType,
+    dim: Int
+  ): VoyagerReader
+  private[voyager] def saveAndClose(voyagerWriter: VoyagerWriter): Unit
+  private[voyager] def exists: Boolean
+}
+
+private[voyager] object VoyagerUri {
+  def apply(path: String, opts: PipelineOptions): VoyagerUri = {
+    if (ScioUtil.isLocalUri(new URI(path))) {
+      new LocalVoyagerUri(path)
+    } else {
+      new RemoteVoyagerUri(path, opts)
+    }
+  }
+
+  def files: Seq[String] = Seq("index.hnsw", "names.json")
+  implicit val voyagerUriCoder: Coder[VoyagerUri] = Coder.kryo[VoyagerUri]
```

Comment:
we could probably do better than `kryo` if `VoyagerUri` is a `sealed trait` and implementations `case classes`

Reply:
I'm curious if you could follow up on this with more of an example?

Reply:
sure! basically, Scio can leverage Magnolify to automatically derive compile-time Coders for Scala ADTs: case classes, and sealed traits (`sealed` is important so that Magnolify knows in advance every possibility it can expect when it decodes an instance of the trait). These types of coders are much more efficient than Kryo, which basically traverses the class fields at runtime.

So, you could have a setup like this:

```scala
sealed trait VoyagerUri {
   def path: String
   ...
}

case class LocalVoyagerUri(path: String) extends VoyagerUri {
   override def....
}

case class RemoteVoyagerUri(path: String, private rfu: RemoteFileUtil) extends VoyagerUri {
   override def ...
}

object RemoteVoyagerUri {
  def apply(path: String, options: PipelineOptions): RemoteVoyagerUri =
   RemoteVoyagerUri(path, RemoteFileUtil.create(options))
}
```

Then you can remove the implicit Kryo coder, and it should just work ™️ .

We even have a specialized type of unit test assertion for checking that your Coder doesn't fall back to kryo:

```scala
import com.spotify.scio.testing.CoderAssertions._

class VoyagerTest extends PipelineSpec {
   "VoyagerUri" should "not use Kryo" in {
      val localUri: VoyagerUri = LocalVoyagerUri("blah")
      localUri coderShould notFallback()

      val remoteUri: VoyagerUri = RemoteVoyagerUri("blah",  PipelineOptionsFactory.create())
      remoteUri coderShould notFallback()
   }
}
```

(more examples in [CoderTest](https://github.com/spotify/scio/blob/e23cc63e3016905a06737a8294923c76162fd769/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala#L533)).

Reply:
refactored this so we do not have ADT. `VoyagerUri` is a simple value-class around a normal URI. the `RemoteFileUtil` can be passed as implicit paramer to `exist`

The other functions should not de defined into the model. Moved them to the `side-input` or the dedicated `asVoyager` SCOllection operation